## 3.3  Approximate Solutions and Bisection Search

In [12]:
#exhaustive enumerative square root
x = 123456
epsilon = 0.01
step = epsilon**3
numGuesses = 0
ans = 0.0
while abs(ans**2 - x) >= epsilon and ans*ans <= x:
    ans += step
    numGuesses += 1
print('numGuesses =', numGuesses)
if abs(ans**2 - x) >= epsilon:
    print('Failed on square root of', x)
else:
    print(ans, 'is close to square root of', x)

KeyboardInterrupt: 

This is an example of approximating a square root using exhaustive enumeration. After we try to make it find a close to exact solution, the program now takes too long to run, and if it's faster it doesn't find the solution we want. So, we have to use a different method.

Suppose we know that a good approximation to the square root of `x` lies somewhere between `0` and `max`. We can exploit the fact that numbers are **totally ordered**. That is to say, for any pair of distinct numbers,`n1` and `n2`, either `n1 < n2` or `n1 > n2`. So we can think of the square root of x as lying somewhere on the line  
`0_____________________________________________________max`  
and start searching that interval. Since we don't necessarily know where to start searching, let's start in the middle.  
`0_________________________guess_______________________max`  
If that's the right answer(most of the time it won't be), ask whether it is too big or too small. If too big, we know the answer must lie to the left. If too small, the answer must be to the right. Then we repeat the process on the smaller interval. This is called a **bisection search**. Here's an example, approximating a square root below.

In [22]:
#Figure 3.4 Using a bisection search to approximate square root
x = 25
epsilon = 0.01
numGuesses = 0
low = 0.0
high = max(1.0, x)
ans = (high + low)/2.0
while abs(ans**2 - x) >= epsilon:
    print('low =', low, 'high =', high, 'ans =', ans)
    numGuesses += 1
    if ans**2 < x:
        low = ans
    else:
        high = ans
    ans = (high + low)/2.0
print('numGuesses =', numGuesses)
print(ans, 'is close to square root of', x)

low = 0.0 high = 25 ans = 12.5
low = 0.0 high = 12.5 ans = 6.25
low = 0.0 high = 6.25 ans = 3.125
low = 3.125 high = 6.25 ans = 4.6875
low = 4.6875 high = 6.25 ans = 5.46875
low = 4.6875 high = 5.46875 ans = 5.078125
low = 4.6875 high = 5.078125 ans = 4.8828125
low = 4.8828125 high = 5.078125 ans = 4.98046875
low = 4.98046875 high = 5.078125 ans = 5.029296875
low = 4.98046875 high = 5.029296875 ans = 5.0048828125
low = 4.98046875 high = 5.0048828125 ans = 4.99267578125
low = 4.99267578125 high = 5.0048828125 ans = 4.998779296875
low = 4.998779296875 high = 5.0048828125 ans = 5.0018310546875
numGuesses = 13
5.00030517578125 is close to square root of 25


Bisection search is a huge improvement over our earlier algorithm, which reduced the search space by only a small amount at each iteration. Bisection search divides the search space in half at each step.

**Finger exercise:** What would the code in Figure 3.4 do if the statement `x = 25` were replaced by `x = -25`?  
~~The code would end really fast because max would be 1.0 instead of 25, and low would still be 0.~~ Ok, after testing it, the code produces an infinite loop. I suppose this is the case because x = -25, but low = 0 and high = 1, so the ans will be a number approaching zero, and the while loop will go on forever because `ans**2 - x` will never not be $\geq$ to `epsilon`, and it will never find the real answer.  

**Finger exercise:** What would have to be changed to make the code in Figure 3.4 work for finding an approximation to the cube root of both negative and positive numbers? (Hint: think about changing `low` to ensure that the answer lies within the region being searched.)  
Well this was simple, all we had to do was change `ans**2` to `ans**3`, and then change the range if x was negative.

In [20]:
#Finger exercise to approximate the cube root of a positive or negative number
x = -27
epsilon = 0.01
numGuesses = 0
#sets range based on if x is positive or negative
if x * -1 == abs(x):
    low = x
    high = 0
else:
    low = 0.0
    high = max(1.0, x)
ans = (high + low)/2.0
while abs(ans**3 - x) >= epsilon:
    print('low =', low, 'high =', high, 'ans =', ans)
    numGuesses += 1
    if ans**3 < x:
        low = ans
    else:
        high = ans
    ans = (high + low)/2.0
print('numGuesses =', numGuesses)
print(ans, 'is close to cube root of', x)

low = -27 high = 0 ans = -13.5
low = -13.5 high = 0 ans = -6.75
low = -6.75 high = 0 ans = -3.375
low = -3.375 high = 0 ans = -1.6875
low = -3.375 high = -1.6875 ans = -2.53125
low = -3.375 high = -2.53125 ans = -2.953125
low = -3.375 high = -2.953125 ans = -3.1640625
low = -3.1640625 high = -2.953125 ans = -3.05859375
low = -3.05859375 high = -2.953125 ans = -3.005859375
low = -3.005859375 high = -2.953125 ans = -2.9794921875
low = -3.005859375 high = -2.9794921875 ans = -2.99267578125
low = -3.005859375 high = -2.99267578125 ans = -2.999267578125
low = -3.005859375 high = -2.999267578125 ans = -3.0025634765625
low = -3.0025634765625 high = -2.999267578125 ans = -3.00091552734375
numGuesses = 14
-3.000091552734375 is close to cube root of -27


***
## 3.4 A Few Words About Using Floats

In [4]:
x = 0.0
for i in range(10):
    x = x + 0.1
if x == 1.0:
    print(x, '= 1.0')
else:
    print(x, 'is not 1.0')

0.9999999999999999 is not 1.0


A binary number is represented by a sequence of digits each of which is either 0 or 1. These digits are often called **bits**. The rightmost digit is the $2^0$ place, the next digit towards the left is the $2^1$ place, etc. For example, the sequence of binary digits `101` represents $1*4 + 0*2 + 1*1 = 5.$ How many different numbers can be represented by a sequence of length $n?\: 2^n.$

**Finger exercise**: What is the decimal equivalent of the binary number `10011`? $$2^0 * 1 + 2^1 * 1 + 2^2 * 0 + 2^3 * 0 + 2^4 * 1$$
$$= 1 + 2 + 0 + 0 + 16 = 19$$

- Non-integer numbers are implmeented using a represtantation called **floating point**.
- Represent a number(in decimal) as a pair of integers, the **significant digits** of the number and an **exponent**.
    - For example, 1.949 would be represented as the pair (1949, -3), which stands for the product $1949 \times 10^{-3}.$
- The number of significant digits determines the **precision** with which numbers can be represented.
    - For example, if there were only two significant digits, the number `1.949` could not be represented exactly. It would have to be converted to an approximation, in this case `1.9`.
- That approximation is called the **rounded value**.

**Binary representations**
- Modern computers use binary, not decimal, representations.
- For example, the number `0.625 (5/8)` would be represented by the pair `(101,011)`; because `5/8` is `0.101` in binary and `-11` is the binary representation of `-3`, the pair `(101,-11)` stands for $5 \times 2^{-3} = 5/8 = 0.625.$

In [4]:
x = 0.0
for i in range(10):
    x = x + 0.1
if x == 1.0:
    print(x, '= 1.0')
else:
    print(x, 'is not 1.0')

0.9999999999999999 is not 1.0


So going back to this example, well in my interpreter we can see that the output is not exactly equal to 1.0, so it returns the output that it does.
- Adding 0.1 ten times does not produce the same result as multiplying 0.1 by 10 in the world of computers.
- Print x will print 1.0 rather than the variable x because Python does some automatic rounding.
- However, what's being displayed does not necessarily match the value stored in the machine.

If you want to explicitly round a floating point number, use the `round` function. The expression `round(x, numDigits)` returns the floating point number equivalent to rounding the value of `x` to `numDigits` decimal digits following the decimal point.

In [5]:
#For example:
print(round(2**0.5,3))

1.414


Most of the time the difference between floating point and real numbers doesn't matter too much. However, one thing that is almost always worth worrying about is tests for equality.  
**It is almost always more appropriate to ask whether two floating point values are close enough to each other, not whether they are identical.**  
For example, it is better to write `abs(x-y) < 0.0001` rather than `x ==y`.

In [6]:
# I guess this value is the epsilon we've seen floating around.

Another thing to worry about is the accumulation of rounding errors. Most of the time things work out OK because sometimes the number stored in the computer is a little bigger than intended, and sometimes it is a little smaller than intended. However, in some programs, the errors will all be in the same direction and accumulate over time.

In [7]:
# This one just got me, in the guess my number bisection search exercise.

***
## 3.5 Newton-Raphson

The most commonly used approximation algorithm is usually attributed to Isaac Newton. It's typically called Newton's method, but sometimes it's called the Newton-Raphson method. It can be used to find the real roots of monay functions, but we're looking at it in the context of finding the real roots of a polynomial with one variable.

A **polynomial** with one variable(by convention, we will write the variable as `x`) is either zero or the sum of a finite number of nonzero terms, e.g., $3x^2 + 2x + 3.$ Each term consists of a constant(the **coefficient** of the term) multiplied by a variable, raised to a nonnegative integer exponent(**degree**). The degree of a polynomial is the largest degree of any single term.

If $p$ is a polynomial and $r$ a real number, we will write $p(r)$ to stand for the value of the polynomial when $x = r.$ A **root** of the polynomial $p$ is a solution to the equation $p = 0$, i.e., an $r$ such that $p(r) = 0.$ So, for example, the problem of finding an approximation to the square root of `24` can be formulated as finding an $x$ such that $x^2 - 24 \approx 0.$

Newton proved a theorem that implies that if a value, call it `guess`, is an approximation to a root of a polynomial, then `guess` - p(`guess`)/p'(`guess`), where `p'` is the first derivative of `p`, is a better approximation.

For any constant $k$ and any coefficient $c$, the first deriviative of $cx^2 + k$ is $2cx.$ For example, the first derivative of $x^2 - k$ is $2x.$ Therefore, we know that we can improve on the current guess, call it $y$, by choosing as our next guess $y - (y^2 - k)/2y.$ This is called **successive approximation**. 

In [11]:
# Newton-Raphson for square root
# Find x such that x**2 - 24 is within epsilon of 0
epsilon = 0.01
k = 24.0
guess = k/2.0
while abs(guess*guess - k) >= epsilon:
    guess = guess - (((guess**2)) - k) /(2*guess)
print('Square root of', k, 'is about', guess)

Square root of 24.0 is about 4.8989887432139305


**Finger exercise:** Add some code to the implementation of Newton-Raphson that keeps track of the number of iterations used to find the root. use that code as part of program that compares the efficiency of Newton-Raphson and bisection search. (You should discover that Newton-Raphson is more efficient.)

In [29]:
# Newton Bisection Comparison
# Bisection
x = int(input("Input a number: "))
epsilon = 0.01
numGuesses = 0
low = 0.0
high = max(1.0, x)
ans = (high + low)/2.0
while abs(ans**2 - x) >= epsilon:
    numGuesses += 1
    if ans**2 < x:
        low = ans
    else:
        high = ans
    ans = (high + low)/2.0
#print('numGuesses =', numGuesses)
#print(ans, 'is close to square root of', x)
# Newton
epsilon = 0.01
iteration = 0
k = x
guess = k/2.0
while abs(guess*guess - k) >= epsilon:
    guess = guess - (((guess**2)) - k) /(2*guess)
    iteration += 1
#print('Square root of', k, 'is about', guess)
print("The square root of", x, "is", ans, "using bisection, and", guess, 'using Newton.')
print('Bisection went through', numGuesses, 'guesses, and Newton went through', iteration, 'guesses.')
if(iteration < numGuesses):
    print('Newton is faster than Bisection.')
else:
    print('Bisection is faster than Newton.')

Input a number: 50
The square root of 50 is 7.0709228515625 using bisection, and 7.0710679289844185 using Newton.
Bisection went through 13 guesses, and Newton went through 5 guesses.
Newton is faster than Bisection.


# 4 Functions, Scoping, And Abstraction

- We've introduced numbers, assignments, input/output, comparisons, and looping constructs.
- In a theoretical sense, this is as powerful as you will ever need.
- Such languages are called **Turing complete.** This means if a problem can be solved by computation, it can be solved by only these operations that we've done.
- All of our code has been a single sequence of instructions, all merged together.

In [ ]:
#Figure 3.4 Using a bisection search to approximate square root
x = 25
epsilon = 0.01
numGuesses = 0
low = 0.0
high = max(1.0, x)
ans = (high + low)/2.0
while abs(ans**2 - x) >= epsilon:
    print('low =', low, 'high =', high, 'ans =', ans)
    numGuesses += 1
    if ans**2 < x:
        low = ans
    else:
        high = ans
    ans = (high + low)/2.0
print('numGuesses =', numGuesses)
print(ans, 'is close to square root of', x)

- This is a reasonable piece of code, but we aren't able to reuse it because it only works for the values denoted by `x` and `epsilon`. So, we can't use it other more complicated computations.
- If we wanted a program to compute cube roots, or a program that computes square roots and cube roots, or even a program that computes square roots in two different places, the program would contain chunks of almost identical code. 
- This is very bad.
- **The more code a program contains, the more chance there is for something to go wrong, and the harder the code is to maintain.**  

- Python provides several linguistic features that make it relatively easy to generalize and reuse code.
- The most important is the function.
***

## 4.1 Functions and Scoping

### 4.1.1 Function Definitions 

In python each **function definition** is of the form  

In [34]:
# def name of function (list of formal parameters):
#     body of function

For example, we could define the function `max` by the code

In [35]:
def max(x,y):
    if x > y:
        return x
    else:
        return y

- `def` is a reserved word that tells Python that a function is about to be defined. The function name is simply a name that is used to refer to the function.
- The names in the parentheses following the function name are the **formal parameters** of the function.
- When the function is used, the formal parameters are bound(as in an assighment statement) to the **actual parameters**(often referred to as **arguments**) of the **function invocation**(also refered to as a **function call**).
* For example, `max(3,4)` binds 3 to x, and 4 to y.

- **return** is a special statement that can only be used within the body of a function
- A function call is an expression, and like all expressions it has a value.
- That value is returned by the invoked function.

To summarize, when a function is called:
1. The expression that make up the actual parameters are evaluated, and the formal parameters of the function are bound to the resulting values. For example, the invocation `max(3+4, z)` will bind the formal parameter `x` to `7` and the formal parameter `y` to whatever value the variable `z` has when the invocation is evaluated.
2. The **point of execution** (the next instruction to be executed) moves from the point of invocation to the first statement in the body of the function.
3. The code in the body of the function is executed until either a return statement is encountered, in which case the value of the expression following the `return` becomes the value of the function invocation, or there are no more statements to execute, in which case the function returns the value `None`. (If no expression follows the `return`, the value of the invocation is `None`.)
4. The value of invocation is the returned value.
5. The point of execution is transferred back to the code immediately following the invocation.

Parameters provide something called **lambda abstraction**, allowing programmers to write code that manipulates not specific objects, but instead whatever objects the caller of the function chooses to use as actual parameters.

**Finger exercise**: Write a function `isIn` that accepts two strings as arguments and returns `True` if either string occurs anywhere in the the other, and `False` otherwise. Hint: you might want to use the built-in `str` operation `in`.

In [37]:
def isIn(str1, str2):
    if str1 in str2:
        return True
    elif str2 in str1:
        return True
    else:
        return False

In [40]:
a = input("Enter string 1: ")
b = input("Enter string 2: ")
isIn(a, b)

Enter string 1: samantha
Enter string 2: sam


True

In [41]:
#Sick, that was pretty easy.

### 4.1.2 Keyword Arguments and Default Values

In Python, there are two ways that formal parameters get bound to actual parameters. The most common method, which is the only one we have used thus far, is called **positional**- the first formal parameter is bound to the first actual parameter, the second formal to the second actual, etc. Python also supports what it calls **keyword arguments**, in which formals are bound to actuals using the name of the formal parameter.

In [207]:
#Figure 4.2 Function that prints a name.
def printName(firstName, lastName, reverse):
    if reverse:
        print(lastName + ', ' + firstName)
    else:
        print(firstName, lastName)

This function `printName` assumes that `firstName` and `lastName` are strings and that `reverse` is a Boolean.
- Keyword arguments are commonly usd in conjunction with **default parameter values.** We can, for example, write

In [210]:
def printName(firstName, lastName, reverse = False):
    if reverse:
        print(lastName + ', ' + firstName)
    else:
        print(firstName, lastName)

Default values allow programmers to call a function with fewer than the specified number of arguments. For example

In [211]:
printName('Olga', 'Puchmajerova')
printName('Olga', 'Puchmajerova', True)
printName('Olga', 'Puchmajerova', reverse = True)

Olga Puchmajerova
Puchmajerova, Olga
Puchmajerova, Olga


### 4.1.3 Scoping  


In [214]:
def f(x): #name x used as formal parameter
    y = 1
    x = x + y
    print('x =', x)
    return x

x = 3
y = 2
z = f(x) #value of x used as actual parameter
print('z =', z)
print('x =', x)
print('y =', y)

x = 4
z = 4
x = 3
y = 2


What's going on here? At the call of `f`, the formal parameter `x` is locally bound to the value of the actual parameter `x`. It is important to note that **though the actual and formal parameters have the same name, they are not the same variable.**

Each function defines a new **name space**, also called a **scope**. The formal parameter `x` and the **local variable** `y` that are used in `f` exist only within the scope of the definition of `f`. The assignment statement `x = x + y` within the function body binds the local name `x` to the object `4`. The assignments in `f` have no effect at all on the bindings of the names `x` and `y` that exist outside the scope of `f`. 

Here's one way to think about this:
- At top level, i.e., the level of the shell, a **symbol table** keeps track of all names defined at that level and their current bindings.
- When a function is called, a new symbol table (sometimes called a **stack frame**) is created. This table keeps track of all names defined within the function (including the formal parameters) and their current bindings. If a function is a called from within the function body, yet another stack frame is created.
- When the function completes, its stack frame goes away.

In python, one can always determine the scope of a name by looking at the program text. This is called **static** or **lexical scoping**. Figure 4.3 contains a slightly more elaborate example.

In [ ]:
#Figure 4.3 Nested scope
def f(x):
    def g():
        x = 'abc'
        print('x =', x)
    def h():
        z = x
        print('z =', z)
    x = x + 1
    print('x =', x)
    h()
    g()
    print('x =', x)
    return g

x = 3
z = f(x)
print('x =', x)
print('z =', z)
z()

<img src= "https://puu.sh/BuBqg/31651656f1.png">

- The first column contains the set of names known outside the body of the function `f`, i.e., the variables `x` and `z`, and the function name `f`.
- First assignment statement binds `x` to `3`.
- z = f(x) invokes the function `f` to which `x` is bound. It creates a stack frame.
- We never remove frames from the middle of the stack, but only the most recently added frame.
- **stack**s have a "last in first out" behavior which is why they are named so.
- come back here later to understand stacks better, no need to understand it all now.

### 4.2 Specifications

- Figure 4.5 generalizes the bisection search we used to find square roots in Figure 4.1. It also contains `testFindRoot`, which can be used to test if `findRoot` works as intended.  
- `testFindRoot` is almost as long as `findRoot`.
- **test functions** might seem like a waste of time, but writing testing code can often pay big dividends. It's certainly better than **debugging**.
- The text between triple quotation marks is called a **docstring** in Python. These are used to provide specifications of functions.
- These docstrings can be accessed using the built-in function `help`.


In [5]:
help(abs)

Help on built-in function abs in module builtins:

abs(x, /)
    Return the absolute value of the argument.



In [8]:
#Figure 4.5 Finding an approximation to a root
def findRoot(x, power, epsilon):
    """Assumes x and epsilon int or float, power an int,
        epsilon > 0 & power >= 1
    Returns float y such that y**power is within epsilon of x.
        If such a float does not exist, it returns None"""
    if x < 0 and power%2 == 0:
        return None
    low = min(-1.0, x)
    high = max(1.0, x)
    ans = (high + low)/ 2.0
    while abs(ans**power - x) >= epsilon:
        if ans**power < x:
            low = ans
        else:
            high = ans
        ans = (high + low)/2.0
    return ans

def testFindRoot():
    epsilon = 0.0001
    for x in (0.25, -0.25, 2, -2, 8, -8):
        for power in range(1, 4):
            print('Testing x = ' + str(x) + \
                  ' and power = ' + str(power))
            result = findRoot(x, power, epsilon)
            if result == None:
                print('   No root')
            else:
                print('   ', result**power, '~=', x)

In [9]:
help(findRoot)

Help on function findRoot in module __main__:

findRoot(x, power, epsilon)
    Assumes x and epsilon int or float, power an int,
        epsilon > 0 & power >= 1
    Returns float y such that y**power is within epsilon of x.
        If such a float does not exist, it returns None



A **specification** of a function defines a contract between the implementer of a function and those who will be writing programs that use the function. We will refer to users of a function as its **clients**. The contract can be thought of as containing two parts.
1. **Assumptions**: Conditions that must be met by the clients of the function.
2. **Guarantees**: Describe what the function will return to the clients.

Functions are a way of creating computational elements that we can think of as primitives. Just as we have built-in functions `max` and `abs`, we would like to have the equivalent of a built-in function for finding roots and for many other complex operations. Functions faciliate this by providing decomposition and abstraction.  
- **Decomposition** creates structure. It allows us to break a problem into modules that are reasonably self-contained, and that may be reused in different settings.
- **Abstraction** hides detail. It allows us to use a piece of code as if it were a **black box-- that is, something whose interior details we cannot see, don't need to see, and shouldn't even want to see.** The essence of abstraction is preserving information that is relevant in a given context, and forgetting information that is irrelevant in that context.
- Abstraction is all about forgetting. There are a lot of ways to model this, for example, the auditory apparatus of most teenagers.
    - Teenager says: May I borrow the car tonight?
    - Parent says: Yes, but be back before midnight, and make sure that the gas tank is full.
    - Teenager hears: Yes.
- The teenager has ignored all of the pesky details that he or she considers irrelevant.
***

## 4.3 Recursion

As a descriptive method recursion is widely used, even by people who would never dream of writing a program.  

Consider part of the legal code of the Unisted States defining the notion of a "natural-born" citizen. Roughly speaking, the definition is as follows
1. Any child born inside the United States,
2. Any child born in wedlock outside the United States both of whose parents are citizens of the U.S., as long as one parent has lived in the U.S. prior to the birth of the child and 
3. Any child born in wedlock outside the United States one of whose parents is a U.S. citizen who has lived at least five years in the U.S. prior to the birth of the child, provided that at least two of those years were after the citizen's fourteen birthday.

The first part is simple; if you are born in the United States, you are a natural-born citizen (such as Barack Obama). If you are not born in the U.S., then one has to decide if your parents are U.S. citizens (either natural born or naturalized). To determine if your parents are U.S. citizens, you might have to look at your grandparents, and so on.

In general, a recursive defintion is made up of two parts. There is at least one **base case** that directly specifies the result for a special case (case 1 in the example above), and there is at least one **recursive (inductive) case** (cases 2 and 3 in the example above) that defines the answer in terms of the answer to the question on some other input, typically a simpler version of the same problem.

The world's simplest recursive definition is probably the factorial function (typically written in mathematics using !) on natural numbers. The classic **inductive definition** is,
$$1! = 1\\ (n+1)! = (n+1)*n!$$  
The first equation defines the base case. The second equation defines factorial for all natural numbers, except the base case, in terms of the factorial of the previous number. 

Figure 4.6 contains both an iterative(`factI`) and a recursive(`factR`) implementation of factorial. 

In [10]:
#Figure 4.6 Iterative and recursive implementation of factorial
def factI(n):
    """Assumes that n is an int > 0
        Returns n!"""
    result = 1
    while n > 1:
        result = result * n
        n -= 1
    return result

def factR(n):
    """Assumes that n is an int > 0
        Returns n!"""
    if n == 1:
        return n
    else:
        return n*factR(n - 1)

This function is sufficiently simple that neither implmeentation is hard to follow. Still, the second is a more obvious translation of the original recursive definition.  
It almost seems like cheating to implement `factR` by calling `factR` from within the body of `factR`. It works for the same reason that the iterative implementation works. We know that the iteration of `factI` will terminate because `n` starts out positive and each time around the loop it is reduced by 1. This means that it cannot be greater than 1 forever. Similarly, if `factR` is called with `1`, it returns a value without making a recursive call. When it does make a recursive call, it always does so with a value one less than the value with which it was called. Eventually the recursion terminates with the call `factR(1)`.

### 4.3.1 Fibonacci Numbers  

The Fibonacci sequence is another common mathematical function that is usually defined recursively. Fibonacci developed this formula to quantify the growth of a population.  

Suppose a newly born pair of rabbits, one male and one female are:
- put into a pen/released into the wild
- able to mate at the age of one month
- have a gestation period of one month
- never die
- the female alawys produces a new pair(male and female) every month from its second month on  
How many pregnant rabbits will there be at the end of six months?

| Month  | Females  |
|---|---|
| 0  | 1  |
| 1  | 1  |
| 2  | 2  |
| 3  | 3  |
| 4  | 5  |
| 5  | 8  |
| 6  | 13  |  

Notice that for month $n > 1,\: \text{females}(n) = \text{females}(n-1) + \text{females}(n-2).$ This is not an accident. Each female that was alive in month `n-1` will still be alive in the month `n`. The new females can be added to the females alive in month `n-1` to get the number of females in month `n`.

The growth in population is described naturally by the **recurrence**:
```Python
females(0) = 1
females(1) = 1
females(n + 2) = females(n+1) + females(n)
```

This definition is a little different from the recursive definition of factorial:
- It has two base cases, not just one. **In general, you can have as many base cases as you need**.
- In the recursive case, there are **two recursive calls, not just one. Again, there can be as many as you need.**

In [16]:
#Figure 4.7 Recursive implementation of Fibonacci sequence
def fib(n):
    """Assumes n an int >= 0
        Returns Fibonacci of n"""
    if n == 0 or n == 1:
        return 1
    else:
        return fib(n-1) + fib(n-2)

def testFib(n):
    for i in range(n+1):
        print('fib of', i, '=', fib(i))

**Finger exercise:** When the implementation of `fib` in Figure 4.7 is used to compute `fib(5)`, how many times does it compute the value of `fib(2)`?

$fib(5)\\
= fib(5-1) + fib(5-2)\\
= fib(4) + fib(3)\\
= fib(4-1) + fib(4-2) + fib(3-1) + fib(3-2)\\
= fib(3) + fib(2) + fib(2) + fib(1)\\
= fib(3-1) + fib(3-2) + 2(fib(2-2) + fib(2-1)) + 1\\
= fib(2) + fib(1) + 2(fib(0) + fib(1)) + 1\\
= fib(2-2) + fib(2-1) + 1 + 2(2) + 1\\
= 2 + 1 + 4 + 1\\
= 8
$

In [17]:
fib(5)

8

I guess it computes `fib(2)` 3 times?

### 4.3.2 Palindromes

Recursion is also useful for many problems that do not involve numbers. Figure 4.8 contains a function, `isPalindrome`, that checks whether a string reads the same way backwards and forwards.

```Python
def isPalindrome(s):
    """Assumes s is a str
        Returns True if the letters in s form a palindrome;
            False otherwise. Non-letters and capitalization are ignored."""
    
    def toChars(s):
        s = s.lower()
        letters = ''
        for c in s:
            if c in 'abcdefghijklmnopqrstuvwxyz':
                letters = letters + c
        return letters
    
    def isPal(s):
        if len(s) <= 1:
            return True
        else:
            return s[0] == s[-1] and isPal(s[1:-1])
        
    return isPal(toChars(s))
```
**<p style="text-align: center;"> Figure 4.8 Palindrome testing </p>**

- The function `isPalindrome` contains two **helper functions**.
- The helper function `toChars` converts all letters to lowercase and removes all non-letters.
- s.lower() is a built-in method that creates a string identical to s, but all the uppercase letters are converted to lowercase letters.
- **method invocation** will be covered more when we get to classes. For now think of it like peculiar syntax for a function call. Instead of putting the first(and in this case only) argument inside parentheses following the function name, we use **dot notation** to place that argument before the function name.
- The helper function `isPal` uses recursion to do the real work.
- The two base cases are strings of lenght zero or one.
- This means that the recursive part of the implementation is reached only on strings of length two or more.
- The conjunction in the `else` clause is evaluated from left to right. The code first checks whether the first and last characters are the same, and if they are goes on to check whether the string minus those two characters is a palindrome.
- When two Boolean-valued expressions are connected by "and," each expression is called a **conjunct**. If they are connected by "or," they are called **disjuncts**.
- The code first checks whether the first and last characters are the same, and if they are, goes on to check whether the string minus those two characters is a palindrome.

The implementation of `isPalindrome` is an example of a problem-solving principle known as **divide-and-conquer**. (This principle is related but different from divide-and-conquer algorithms.) The problem-solving principle is to conquer a hard problem by breaking it into a set of subproblems with the properties that
- the subproblems are easier to solve than the original problem, and
- solutions of the subproblems can be combined to solve the original problem.  
In this case, we solve the problem by breaking the original problem into a simpler version of the same problem (checking whether a shorter string is a palindrome), plus some simple things we know how to do (comparing single characters). Figure 4.9 contains some code that can be used to visualize how this works.

***
## 4.4 Global Variables

If you tried called `fib` with a large number, you probably noticed that it took a very long time to run. Suppose we want to know how many recursive calls are made? We could do a careful analysis of the code and figure it out, and in Chapter 9 we will talk about how to do that. Another approach is to add some code that counts the number of calls. One way to do that uses **global variables**.

```Python
def fib(x):
    """Assumes x an int >= 0
       Returns Fibonacci of x"""
    global numFibCalls
    numFibCalls += 1
    if x == 0 or x == 1:
        return 1
    else:
        return fib(x-1) + fib(x-2)
    
def testFib(n):
    for i in range(n+1):
        global numFibCalls
        numFibCalls = 0
        print('fib of', i, '=', fib(i))
        print('fib called', numFibCalls, 'times.')
```
**<p style="text-align: center;"> Figure 4.10 Using a global variable </p>**

In each function, the line of code `global numFibCalls` tells Python that the name `numCalls` should be defined at the outermost scope of the module in which the line of code appears rather than within the scope of the function in which the line of code appears--despite the fact that `numFibCalls` occurs on the left-hand side of an assignment statement in both `fib` and `testFib`. (Had we not included the code `global numFibCalls`, the name `numFibCalls` would have been local to each of `fib` and `testFib`.) The functions `fib` and `testFib` both have unfettered access to the object referenced by the variable `numFibCalls`. The function `testFib` binds `numFibCalls` to 0 each time it calls `fib`, and `fib` increments the value of `numFibCalls` each time `fib` is entered.

***
## 4.5 Modules

So far, we have operated under the assumption that our entire program is stored in one file. This is perfectly reasonable as long as programs are small. As programs get larger, however, it is typically more convenient to store different parts of them in different files. Imagine, for example, that multiple people are working on the same program. It would be a nightmare if they were all trying to update the same file. Python modules allow us to easily construct a program from code in multiple files.

A **module** is a `.py` file containing Pythong definitions and statements. We could create, for example, a file `circle.py` containing
```Python
pi = 3.14159

def area(radius):
    return pi*(radius**2)

def circumference(radius):
    return 2*pi*radius

def sphereSurface(radius):
    return 4.0*area(radius)

def sphereVolume(radius):
    return (4.0/3.0)*pi*(radius**3)
```

A program gets access to a module through an `import` statement. So, for example, the code
```Python
    import circle
    print(circle.pi)
    print(circle.area(3))
    print(circle.circumference(3))
    print(circle.sphereSurface(3))
```
will print
```Python
    3.14159
    28.27431
    18.84954
    113.09724
```

Modules are typically stored in individual files. Each module has its own private symbol area. Consequently, within `circle.py` we access objects (e.g. `pi` and `area`) in the usual way. Executing `import M` creates binding for `M` in the scope in which the importation occurs. Therefore, in the importing context we use dot notation to indicate that we are referring to a name defined in the imported module. For example, outside of `circle.py`, the references `pi` and `circle.pi` can (and in this case do) refe to different objects.

At first glance, the use of dot notation may seem cumbersome. On the other hand, when one imports a module one often has no idea what local names might have been used in the implmentation of that module. The use of dot notation to fully qualify names avoids the possibility of getting burned by an accidental name clash. For example, the assignment statement `pi = 3.0` does not change the value of `pi` used within the `circle` module.

There is a variant of the `import` statement that allows the importing program to omit the module name when accessing names defined inside the imported module. Executing the statement `from M import *` creates bindings in the current scope to all objects defined within `M`, but not to `M` itself. For example, the code
```Python
    from circle import *
    print(pi)
    print(circle.pi)
```  
will first print `3.14159` and then produce the error message  
```Python
    NameError: name `circle` is not defined
```  
Some Python programmers frown about using this form of `import` because they believe that it makes code more difficult to read.  

- A module is imported only once per interpreter session. If you start up IDLE, import a module, and then change the contents of that module, the interpreter will still be using the original version of the module. 
- You can force the interpreter to reload all imported modules by executing `reload()`.  
***

## 4.6 Files

Each operating system comes with its own file system for creating and accessing files. Python achieves operating-system independence by accessing files through something called a **file handle**. The code  
    `nameHandle = open('kids', 'w')`  
instructs the operating system to create a file with the name `kids`, and return a file handle for that file. The argument `w` to `open` indicates that the file is to be opened for writing. The following code opens a file, uses the **write** method to write two lines, and then clsoes the file. It is important to remember to close the file when the program is finished using it. Otherwise there is a risk that some or all of the writes may not be saved.  
```Python
    nameHandle = open('kids', 'w')
    for i in range(2):
        name = raw_input('Enter name: ')
        nameHandle.write(name + '\n')
    nameHandle.close()
```  
In a string, the character "\" is an escape character used to indicate that the next character should be treated in a special way. In this example, the string '\n' indicates a new line character.  
We can now open the file for **reading**(using the argument 'r'), and print its contents. Since Python treats a file as asequence of lines, we can use a `for` statement to iterate over the file's contents.  
```Python
    nameHandle = open('kids', 'r')
    for line in nameHandle:
        print(line)
    nameHandle.close()
```  
If we had typed in the names David and Andrea, this will print
```Python
    David
    
    Andrea
```  
The extra line between DAvid and Andrea is there because print starts a new line each time it encoutners the '/n' at the end of each line in the file. We could have avoided printing that by writing `print(line[:-1])`. Now consider  
```Python
    nameHandle = open('kids', 'w')
    nameHandle.write('Michael\n')
    nameHandle.write('Mark\n')
    nameHandle.close()
    nameHandle = open('kids', 'r')
    for line in nameHandle:
        print(line[:-1])
    nameHandle.close()
```  
It will print  
```Python
    Michael
    Mark
```  
Notice that we have overwritten the previous contents of the file `kids`. If we don't want to do that we can open the file for **appending** (instead of writing) by using the argument 'a'.

Ok we're done, we didn't really understand that much though tbh.